substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/PosixJavaNetSubstitutions.java

```diff
@@ -119,84 +419,1339 @@ protected int peekData(DatagramPacket p) throws IOException {
+
+
+        //   957      jobject fdObj = (*env)->GetObjectField(env, this, pdsi_fdID);
+        FileDescriptor fdObj = Util_java_net_DatagramSocketImpl.as_Target_java_net_DatagramSocketImpl(this).fd;
+        //   958      int arg, fd, t = 1;
+        int arg, fd, t = 1;
+
+        int domain;
+        //   960  #ifdef AF_INET6
+        if (IsDefined.socket_AF_INET6()) {
+            //   961      int domain = ipv6_available() ? AF_INET6 : AF_INET;
+            domain = JavaNetNetUtil.ipv6_available() ? Socket.AF_INET6() : Socket.AF_INET();
+            //   962  #else
+        } else {
+            //   963      int domain = AF_INET;
+            domain = Socket.AF_INET();
+            //   964  #endif
+        }
+
+        //   966      if (IS_NULL(fdObj)) {
+        if (fdObj == null) {
+            //   967          JNU_ThrowByName(env, JNU_JAVANETPKG "SocketException",
+            //   968                          "Socket closed");
+            //   969          return;
+            throw new SocketException("Socket closed");
+            //   970      }
+        }
+
+        //   972      if ((fd = JVM_Socket(domain, SOCK_DGRAM, 0)) == JVM_IO_ERR) {
+        if ((fd = Socket.socket(domain, Socket.SOCK_DGRAM(), 0)) == Target_jvm.JVM_IO_ERR()) {
+            //   973          NET_ThrowByNameWithLastError(env, JNU_JAVANETPKG "SocketException",
+            //   974                         "Error creating socket");
+            throw new SocketException(PosixUtils.lastErrorString("Error creating socket") + " (Error creating socket)");
+            //   975          return;
+            /* Unreachable. */
+        }
+
+        //   978  #ifdef AF_INET6
+        if (IsDefined.socket_AF_INET6()) {
+            //   979      /* Disable IPV6_V6ONLY to ensure dual-socket support */
+            //   980      if (domain == AF_INET6) {
+            if (domain == Socket.AF_INET6()) {
+                //   981          arg = 0;
+                arg = 0;
+                CIntPointer arg_Pointer = StackValue.get(CIntPointer.class);
+                arg_Pointer.write(arg);
+                //   982          if (setsockopt(fd, IPPROTO_IPV6, IPV6_V6ONLY, (char*)&arg,
+                //   983                         sizeof(int)) < 0) {
+                if (Socket.setsockopt(fd, NetinetIn.IPPROTO_IPV6(), NetinetIn.IPV6_V6ONLY(), arg_Pointer, SizeOf.get(CIntPointer.class)) < 0) {
+                    //   984              NET_ThrowNew(env, errno, "cannot set IPPROTO_IPV6");
+                    try {
+                        JavaNetNetUtilMD.NET_ThrowNew(Errno.errno(), "cannot set IPPROTO_IPV6");
+                    } finally {
+                        //   985              close(fd);
+                        //   986              return;
+                        Unistd.close(fd);
+                        /* return required to prevent compiler from complaining about unhandled InterruptedIOException */
```

Comment:
(You have `JavaNetNetUtilMD.NET_ThrowNew` throwing an `InterruptedException` instead of an `InterruptedIOException`.  I will try to remember to comment on that when I get to that method.)

Since `PlainDatagramSocketImpl.datagramSocketCreate` is declared to only throw `SocketException`, but we are the only ones who have to live by that rule, I suggest you turn an `InterruptedIOException` plowing through here by throwing a new `SocketException` whose message string is "InterruptedIOException" plus the message string from the `InterruptedIOException`, if that is non-null.

On the other hand, `setsockopt(2)` says it can not return with `errno` set to `EINTR`, so maybe we should not spend too much time agonizing over this.

Reply:
In this case, `NET_ThrowNew(...)` existed prior to my work. I'm just using it.